src/test/scala/chiselTests/TypeAliasSpec.scala

```diff
@@ -0,0 +1,240 @@
+          val out = Output(new BarBundle)
+        })
+
+        io.out.x := io.in.y
+        io.out.y :#= io.in.x
+      }
+
+      val args = Array("--throw-on-first-error", "--full-stacktrace")
+      val chirrtl = ChiselStage.emitCHIRRTL(new Test, args)
+    }).getMessage should include(
+      "Attempted to redeclare an existing type alias 'DifferentBundle' with a new bundle structure"
+    )
+  }
+
+  "Bundles with unsanitary names" should "properly sanitize" in {
+    class Test extends Module {
+      class FooBundle extends Bundle {
+        // Sanitizes to '_'; the sanitized alias needs to be used in the aliasing algorithm
+        // instead of the direct user-defined alias.
+        override def aliasName = Some("")
+
+        val x = UInt(8.W)
+        val y = UInt(8.W)
+      }
+
+      val io = IO(new Bundle {
+        val in = Input(new FooBundle)
+        val out = Output(new FooBundle)
+      })
+
+      val w = Wire(new FooBundle)
+
+      w :#= io.in
+      io.out :#= w
+    }
+
+    val chirrtl = ChiselStage.emitCHIRRTL(new Test)
+
+    chirrtl should include("type _ = { x : UInt<8>, y : UInt<8>}")
+    chirrtl should include("output io : { flip in : _, out : _}")
+    chirrtl should include("wire w : _")
+  }
+
+  "Bundle type aliases overriding an existing FIRRTL type" should "error" in {
+    // Special keywords/types specified in the FIRRTL spec.
+    // These result in parser errors and should not be allowed by Chisel
+    // Duplicated from the same list in the `internal` package object, which is private to chisel3
+    val firrtlKeywords = Seq(
+      "FIRRTL",
+      "Clock",
+      "UInt",
+      "Reset",
+      "AsyncReset",
+      "Analog",
+      "Probe",
+      "RWProbe",
+      "version",
+      "type",
+      "circuit",
+      "parameter",
+      "input",
+      "output",
+      "extmodule",
+      "module",
+      "intmodule",
+      "intrinsic",
+      "defname",
+      "const",
+      "flip",
+      "reg",
+      "smem",
+      "cmem",
+      "mport",
+      "define",
+      "attach",
+      "inst",
+      "of",
+      "reset",
+      "printf",
+      "skip",
+      "node"
+    )
+
+    // Prevent statements like type Clock = { ... }
+    firrtlKeywords.map { tpe =>
+      (the[ChiselException] thrownBy extractCause[ChiselException] {
+        class Test(val firrtlType: String) extends Module {
+          class FooBundle extends Bundle {
+            override def aliasName = Some(firrtlType)
+
+            val x = UInt(8.W)
+          }
+
+          val io = IO(new Bundle {
+            val in = Input(new FooBundle)
+            val out = Output(new FooBundle)
+          })
+
+          io.out.x :#= io.in.x
+        }
+
+        val args = Array("--throw-on-first-error", "--full-stacktrace")
+        val chirrtl = ChiselStage.emitCHIRRTL(new Test(tpe), args)
+      }).getMessage should include(
+        s"Attempted to override a FIRRTL keyword '$tpe' with a type alias!"
```

Comment:
does everything work as expected with various `Input`, `Output` wrapping? Do we expect those to JustWork ,I would think not because i think in the emitted firrtl those `flip` have already been stripped? So do we get the error we might expect if we put a named bundle inside an aggregate parent who has stripped its direction? If this is the case, should we make strpping operations automatically add some sort of standard thing to the name like `MonoDirectioned`?